### PR TITLE
Add completions for Obsidian Callouts

### PR DESCRIPTION
**Pull request by eth-p**
Adds completion/suggestion support for Obsidian [callouts](https://help.obsidian.md/How+to/Use+callouts).

image

- Only triggers when changing the type of a callout.
   ```
   > [!TYPE] Title
      ~~~~~ <-- only triggers for these characters
   ```

- Will show (and correctly autocomplete) when a callout is likely to be created.
   ```
   > [_
      ~ <-- will trigger here
   ```

- Moves cursor in the ideal place to type the title.
   ```
   > [!TY_] _
         |  ~ <-- cursor will move here
         ~ <-- complete from here
   ```

- Correctly handles nested callouts.

## Reviewer Note

Each commit is a distinct change, and is easier to review one-by-one instead of as a whole.

### CODE REVIEW

File: src/provider/callout_provider.ts [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    SuggestionProvider
+} from "./provider";
+import {CompletrSettings} from "../settings";
+import {Notice, Vault} from "obsidian";
+import {SuggestionBlacklist} from "./blacklist";
+
+
+const CALLOUT_TYPES_PATH = ".obsidian/plugins/obsidian-completr/callout_suggestions.json";
+const BLOCKQUOTE_PREFIX_REGEX = /^(?:[ \t]*>[ \t]*)+/;
+const CALLOUT_HEADER_REGEX = /^(\[!?([^\]]*)\])(-?)([ \t]*)(.*)$/d; // [!TYPE]- TITLE
```
Comment (tth05) [inline src/provider/callout_provider.ts:13]: Obsidian also allows a `+` as the foldable argument. See https://help.obsidian.md/How+to/Use+callouts#Folding

File: src/provider/callout_provider.ts [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        // Generate and return the suggestions.
+        const replaceFoldable = callout.foldable.rawText === '' ? ' ' : callout.foldable.rawText;
+        const replaceTitle = callout.title.rawText;
+
+        const search = callout.type.text.toLowerCase();
+        return this.loadedSuggestions
+            .filter(s => s.displayName.toLowerCase().startsWith(search) || s.replacement.toLowerCase().startsWith(search))
+            .map(suggestion => {
+                return suggestion.derive({
+                    replacement: `[!${suggestion.replacement}]${replaceFoldable}${replaceTitle}`,
```
Comment (tth05) [inline src/provider/callout_provider.ts:64]: What's the point of extracting the foldable and title, just to place it back afterwards? Couldn't we instead just replace from the `[` until `]` up to the cursor or the `]`. Then optionally append a `]` or prepend a `!`.
I think this could simplify the regex, the `CalloutHeader` interface and the `extractCalloutHeader` function, as well as the replacement.

**Review by tth05 — CHANGES_REQUESTED** [review]
Looks very nice! Thanks a lot for the contribution

**Comment by tth05** [comment]
On a side note: I was also planning on adding icons, but representing the provider the contributed the given suggestion instead. This would of course conflict with your icons. Do you have an idea how this could be handled? Two icons would look weird, even if one was in the front and one was in the back.
What I'm thinking of right now is maybe the provider icon taking the whole space if the suggestion doesn't have an icon, otherwise it'd be displayed in the corner (`C` for Callout):
image

File: src/provider/callout_provider.ts [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+    Suggestion,
+    SuggestionContext,
+    SuggestionProvider
+} from "./provider";
+import {CompletrSettings} from "../settings";
+import {Notice, Vault} from "obsidian";
+import {SuggestionBlacklist} from "./blacklist";
+
+
+const CALLOUT_TYPES_PATH = ".obsidian/plugins/obsidian-completr/callout_suggestions.json";
```
Comment (tth05) [inline src/provider/callout_provider.ts:11]: If you have the time, it'd be nice if you can rebase and change this path like I did here https://github.com/tth05/obsidian-completr/commit/5e260163f42fdb71835831bcfc71778f8a62151a

File: src/provider/callout_provider.ts (hunk 2 above)
Comment (eth-p) [reply src/provider/callout_provider.ts:65]: It does seem overcomplicated, but it has its reasons :)

In my (subjective) ideal workflow for completing callouts, after the type is entered/completed, the cursor should jump to either the start or end of the title part of the header. It allows me to type `> [exa` <kbd>Tab</kbd> `A Quick Example` to quickly generate a full callout (`> [!example] A Quick Example`) without needing any additional punctuation or using the arrow keys.

From what I could gather, when a completion is accepted, it places the cursor at the end of the now-inserted replacement text. In the above example, that works fine if you don't have the `Editor -. Auto pair brackets` setting enabled. If it is enabled, though, the cursor ends up between the already-existing closing bracket and the completion like this:

```
> [!example|]
``` 

The same issue occurs when working with a callout header that is already filled out, but lacks a title (e.g. `> [!example]+ `).

Another problem I had with that approach was that there's a variable amount of spacing^1 between the foldable option (or lack of) and the title. It was a fair bit of extra logic to reposition the cursor in the correct spot, insert a space if there wasn't one (but only if there was no title text) while also accounting for the foldable flag, etc. It also had to not clobber the existing whitespace (e.g. if the user preferred a tab or multiple spaces, it should keep that). At that point, a bunch of lines were dedicated just to do the offset counting and string concatenation for the replacement text, as well as introducing another parameter for the `Suggestion` that told it how many extra characters the cursor should be moved *after* the replacement was inserted. It was honestly easier and less boilerplate-y to rebuild and replace the whole line.

**In summary:**

It just wasn't worth it to deal with the edge cases when I could instead solve both problems by extracting important details from the and rebuilding/replacing the entire line at once. The cursor lands at the end of title of, the foldable options don't get removed by accident, and user-provided whitespace between the `-` and title text don't get replaced or clobbered.

---

**Footnote 1:** As I was double checking this, I learned there actually has to be at least one character worth of whitespace between the type (or foldable options) and the title, or it wouldn't render as a callout. I added another commit which makes `completr` fix that when the user accepts the completion.

File: src/provider/callout_provider.ts (hunk 3 above)
Comment (eth-p) [reply src/provider/callout_provider.ts]: Done :)

**Comment by eth-p** [comment]
Regarding the side note:

I prefer only seeing the provider when there isn't an icon, as a way to help keep the suggestions all nicely aligned in a grid. But, I also see the value in knowing where the suggestions came from. In terms of usability and aesthetics, I think your idea of overlying the provider on top of the icon in the corner is the best approach, though. At the same time, I'm a fan of configurability and "having it your way", so maybe it could be an option in the settings?

For example:

- image (if the suggestion has an icon, show both. otherwise, show provider icon)
- image (if the suggestion has an icon, show that. otherwise, show provider icon)
- image (always show the provider icon, even if the suggestion has an icon) 
- image (no icons at all. for those who prefer a minimalist look)

File: src/provider/callout_provider.ts (hunk 1 above)
Comment (eth-p) [reply src/provider/callout_provider.ts]: Fixed! Thanks for pointing that out :)

File: src/provider/callout_provider.ts (hunk 2 above)
Comment (tth05) [reply src/provider/callout_provider.ts:65]: Thanks for the explanation. Because you already tried it without this, I'll trust your instinct that this is nicer :^)

**Comment by tth05** [comment]
> Regarding the side note:
> 
> I prefer only seeing the provider when there isn't an icon, as a way to help keep the suggestions all nicely aligned in a grid. But, I also see the value in knowing where the suggestions came from. In terms of usability and aesthetics, I think your idea of overlying the provider on top of the icon in the corner is the best approach, though. At the same time, I'm a fan of configurability and "having it your way", so maybe it could be an option in the settings?

Multiple options seem like the best way. I extracted your images into a [new issue](https://github.com/tth05/obsidian-completr/issues/46).
